### PR TITLE
feat: Ability to create custom labels for namespaces created with syncOptions CreateNamespace 

### LINKED ISSUE #4628: Add the ability to add annotations to all automatically created namespaces
# Summary

Currently Argo CD can automatically create namespaces. We use Rancher UI. These namespaces do not have the RancherUI annotation for Rancher Projects. It would be nice to be able to define on the Project or even application level annotations that get added to automatically created Namespaces.

# Motivation

When using ArgoCD and Rancher UI

# Proposal

I think it could be a project level mapping. That way you could map Argo Projects to Rancher Projects. But possibly with App level overrides. It is really about annotating the namespace.

### LINKED ISSUE #6215: Add default ArgoCD label to auto created namespaces
# Summary

Currently there is no way to see which namespaces were automatically created by ArgoCD. Goal is to add a default label like `argocd.argoproj.io/namespace: production`.

# Motivation

There are a couple of reasons I can honesyly think of:
1. This allows you to easily see which namespaces were created by ArgoCD or manually by a user with a bit to much power (CR/R)
2. You can reference this label in your network policies (namedpace selectors) OOTB, this makes the auto create feature even more powerful
3. In the event you want to empty/clean all namespaces and associated resources with, you can easily use kubectl combined with a label selector

# Proposal

Add a default label like `argocd.argoproj.io/namespace: production` where `production` here would be the actual name of namespace off course.

### LINKED ISSUE #7799: Add ability to create custom labels for namespaces created with syncOptions CreateNamespace
First up let me just say thank you for making such a wonderful tool!

# Summary

At the moment we can configure ArgoCD to create a namespace for us with:

```yaml
syncOptions:
    - "CreateNamespace=true"
 ```

But we can't add custom labels to that namespace.

# Motivation

To achieve zero downtime deployments with the AWS Load Balancer Controller they recommend adding a specific custom label to each namespace you deploy your application into. This is documented at https://kubernetes-sigs.github.io/aws-load-balancer-controller/v2.3/deploy/pod_readiness_gate/. This feels like something you would always want to use on AWS with EKS.

They say to run: `kubectl label namespace readiness elbv2.k8s.aws/pod-readiness-gate-inject=enabled`

Another use case is enabling Fluent Bit on EKS with Fargate, this allows your pods to log to CloudWatch. One requirement is to create a namespace with a custom label. Details are at https://aws.amazon.com/blogs/containers/fluent-bit-for-amazon-eks-on-aws-fargate-is-here/.

In the spirit of gitops you could explicitly create a `namespace.yaml` file and include that as part of your Kubernetes manifest files (raw, Kustomize, Helm, etc.) and then not use Argo's `CreateNamespace=true` but this feels a little dirty I think? It would mean setting a custom overlay with Kustomize for every namespace (prod, staging, qa, pr, etc.), this starts to add up.

If this feature existed we could avoid creating a `namespace.yaml` and instead configure our ArgoCD Application to create a namespace with 0 or more custom labels.

# Proposal

I'm not sure how it should be implemented. Maybe something like:

```yaml
syncOptions:
  namespace:
    create: true
    labels:
      elbv2.k8s.aws/pod-readiness-gate-inject: "enabled"
```

It could default to `labels: {}` to create no labels.

What do you think? In the mean time is the explicit creation of a `namespace.yaml` a reasonable and gitops friendly way to do this?

### PR DESCRIPTION
**Pull request by pasha-codefresh**
Closes https://github.com/argoproj/argo-cd/issues/4628
Closes https://github.com/argoproj/argo-cd/issues/6215
Closes https://github.com/argoproj/argo-cd/issues/7799

### CODE REVIEW

**Comment by todaywasawesome** [comment]
Linked PR in GitOps engine. https://github.com/argoproj/gitops-engine/pull/443

File: pkg/apis/application/v1alpha1/types.go [hunk 1]
```diff
@@ -698,14 +698,20 @@ func (o SyncOptions) HasOption(option string) bool {
 	return false
 }
 
+type CreateNamespaceMetadata struct {
+	Labels      map[string]string `json:"labels,omitempty" protobuf:"bytes,1,opt,name=labels"`
+	Annotations map[string]string `json:"annotations,omitempty" protobuf:"bytes,2,opt,name=annotations"`
```
Comment (leoluz) [inline pkg/apis/application/v1alpha1/types.go:703]: Maybe better providing the name here as well? 
Or even make this struct just have a [ObjectMeta](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.24/#objectmeta-v1-meta) reference in it?

something like:
```golang
type CreateNamespaceMetadata struct {
	metav1.ObjectMeta `json:"metadata"`
```

Comment (crenshaw-dev) [reply pkg/apis/application/v1alpha1/types.go:703]: The name is inferred from the resource referencing the to-be-created namespace.

**Review by leoluz — CHANGES_REQUESTED** [review]
As discussed a few scenarios that need to be addressed and documented:

- ArgoCD needs to correctly maintain the namespace state as defined in the new `CreateNamespaceMetadata`. This means that it needs to handle create, update and deletes. It is basically similar thing that we do during the reconciliation but with the difference that the desired state doesn't come from git in this case.

- Define what will be the behaviour if the user provides the namespace with specific state in git and also defines the Application resource with a different state. What will be the correct behaviour in this case?

- ~There are some use-cases where users want to define the namespace name dynamically. Are we supporting this use-case as part of this PR?~  (I can't find the comment where the user asked about this feature.. Let's leave it out for now for simplification.)

**Comment by crenshaw-dev** [comment]
> There are some use-cases where users want to define the namespace name dynamically. Are we supporting this use-case as part of this PR?

Can you link those use cases? afaik the namespace names should always be taken from the resources in the sync operation.

**Comment by leoluz** [comment]
As discussed with @pre in the original ticket please consider changing the terminology from `CreateNamespaceMetadata` to `ManagedNamespaceMetadata`.
For more details [click here](https://github.com/argoproj/argo-cd/issues/7799#issuecomment-1220984249).

**Comment by rcarre** [comment]
Many thanks for this enhancement.
Today we are facing SharedResourceWarning since 2 applications manage the same namespace to be created in a specific Rancher project by the means of here enclosed manifest. As this namespace is the only shared resource, it is not worth your while to create a specific application for shared resource.
```---
apiVersion: v1
kind: Namespace
metadata:
  annotations:
    field.cattle.io/projectId: ${ARGOCD_ENV_RANCHER_PROJECT_ID}
  name: ${ARGOCD_APP_NAMESPACE}
```

Thanks to this PR, we expect to trigger `syncOptions:  - CreateNamespace=true ` with specific annotations built from env variable
```
      source:
        plugin:
          name: argocd-vault-plugin-envsubst
          env:
             - name: RANCHER_PROJECT_ID
             - value: ${ARGOCD_ENV_RANCHER_DEFAULT_PROJECT_ID}
      syncPolicy:
        syncOptions: 
          - CreateNamespace=true
```

**Comment by crenshaw-dev** [comment]
@pasha-codefresh do you still have time to move this forwards?

**Comment by blakepettersson** [comment]
I'd be happy to carry the torch if @pasha-codefresh doesn't have time before the cutting of 2.5 :smile:

**Comment by crenshaw-dev** [comment]
@blakepettersson that would be awesome! Hoping to cut RC1 next week, but we'll see how multi-source apps goes.  :-)

**Comment by blakepettersson** [comment]
I see that the required changes to implement this was reverted with argoproj/gitops-engine#455, why's that @leoluz @pasha-codefresh @crenshaw-dev?

**Comment by leoluz** [comment]
> I see that the required changes to implement this was reverted with [argoproj/gitops-engine#455](https://github.com/argoproj/gitops-engine/pull/455), why's that @leoluz @pasha-codefresh @crenshaw-dev?

@blakepettersson That was reverted because it was breaking Argo CD build as it was introducing a breaking change.

**Comment by blakepettersson** [comment]
@leoluz can you give some examples on what you mean here?

> Define what will be the behaviour if the user provides the namespace with specific state in git and also defines the Application resource with a different state. What will be the correct behaviour in this case?

**Comment by gazab** [comment]
> @leoluz can you give some examples on what you mean here?
> 
> > Define what will be the behaviour if the user provides the namespace with specific state in git and also defines the Application resource with a different state. What will be the correct behaviour in this case?

I guess he means what should happen if the Application contains a Namespace resource for the managed namespace. Something like this:

```yaml
syncPolicy: 
  managedNamespaceMetadata:
    annotations: 
      annotation1: x1
      annotation2: y1
    labels: 
      label1: z
      label2: v
```

together with:
```yaml
apiVersion: v1
kind: Namespace
metadata:
  name: same-as-application-namespace
  annotations:
    annotation1: x2
    annotation2: y2
    annotation3: new_annotation_value
```

What values should the namespace end up with?

**Comment by leoluz** [comment]
> I guess he means what should happen if the Application contains a Namespace resource for the managed namespace. Something like this:

Exactly. Tks for clarifying that with examples!
I am tempted to say that Argo CD should raise a conflict error in this case.

Another case is, the namespace resource is not provided with the application but already exists in the cluster with conflicting labels/annotations values. We need to define what would be this behaviour and document it.

**Comment by blakepettersson** [comment]
> Exactly. Tks for clarifying that with examples @gazab !
> I am tempted to say that Argo CD should raise a conflict error in this case.

Indeed, thanks @gazab 😄. IMO at the very least there definitely should be some kind of warning if that happens. Currently when the same k8s resource is managed by two separate Argo applications a `SharedResourceWarning` is emitted; it would make sense for something like that to be emitted when the same application is attempting to create the same namespace both through k8s manifests and `managedNamespaceMetadata`.

Another (complementary?) option could be to say "k8s manifest wins" (with a warning), something like

1. Namespace is created with `managedNamespaceMetadata` and gets synced first
2. Then the application manifests gets applied, overwriting whatever has been done with `managedNamespaceMetadata`.

> Another case is, the namespace resource is not provided with the application but already exists in the cluster with conflicting labels/annotations values. We need to define what would be this behaviour and document it.

Today, if an existing k8s resource gets adopted by an ArgoCD application, it would be overwritten by whatever ArgoCD applies right? It would make sense in that case to be consistent and apply the same behaviour for namespaces as well.

**Comment by leoluz** [comment]
> Today, if an existing k8s resource gets adopted by an ArgoCD application, it would be overwritten by whatever ArgoCD applies right? It would make sense in that case to be consistent and apply the same behaviour for namespaces as well.

@blakepettersson the difference is that in this case it is the same actor (Argo CD Controller) conflicting with itself. The managedFields would be updated with the same manager. However, the overriding logic you described doesn't look bad and I wouldn't be opposed to that behaviour as long as it is documented.

**Comment by blakepettersson** [comment]
I might need another pair of eyes to check that I'm not missing something :smile: 

In #10672, I added some E2E test cases to verify that

1. For a namespace that is not originally managed by ArgoCD, the metadata in `managedNamespaceMetadata` gets applied on the preexisting namespace
2. For a namespace which has conflicting metadata in `managedNamespaceMetadata` and as a k8s manifest, the k8s manifest data will overwrite the data which is in `managedNamespaceMetadata` 

I still need to add warnings and/or errors for these conflicts, that's next on my list.

**Comment by blakepettersson** [comment]
> Another (complementary?) option could be to say "k8s manifest wins" (with a warning), something like
>
>    Namespace is created with managedNamespaceMetadata and gets synced first
>    Then the application manifests gets applied, overwriting whatever has been done with managedNamespaceMetadata.

> I still need to add warnings and/or errors for these conflicts, that's next on my list.

The more I'm thinking about this, the more I think we should disallow updates to namespaces given any conflict. My current line of thinking is 

1. Emit a warning when detecting a conflict. At the moment (locally) it's a `SharedResourceWarning` but can be made into something else.
2. Do not sync any changes that would have otherwise happened to the namespace. 

Thoughts on this @leoluz @crenshaw-dev @gazab?

**Comment by crenshaw-dev** [comment]
What's the downside to just silently accepting whatever is in the manifests?

**Comment by leoluz** [comment]
> What's the downside to just silently accepting whatever is in the manifests?

@crenshaw-dev @blakepettersson I don't think we should do that. Existing namespaces might hold crucial labels and annotations that are necessary for the correct operation of user's clusters. Argo CD shouldn't just replace all annotations and labels as it would certainly break some k8s setups.

I think it would be better continue with further discussions in @blakepettersson PR (https://github.com/argoproj/argo-cd/pull/10672)

I am closing this for now.

**Pull request closed without merging**